### PR TITLE
Update to C# 7.3 and use the new tuple syntax

### PR DESCRIPTION
**Pull request by teinarss**
IMO this is long overdue. There are several new  syntax features that we could take advantage of, for more readable code and better performance (ref structs, ref var).

This would be the first PR in a series of PR that would update:
* String interpolation
* out var

### CODE REVIEW

**Comment by phrohdoh** [comment]
@OpenRA/core Are we ready to consider bumping min versions as necessary to merge this?
My understanding is that it'll be _quite some time still_ before OpenRA does this.

**Comment by teinarss** [comment]
> 
> 
> @OpenRA/core Are we ready to consider bumping min versions as necessary to merge this?
> My understanding is that it'll be _quite some time still_ before OpenRA does this.

We have already bumped the mono version: https://github.com/OpenRA/OpenRA/blob/bleed/INSTALL.md and for mono support see https://www.mono-project.com/docs/about-mono/releases/5.16.0/

**Comment by abcdefg30** [comment]
Fwiw this needs a rebase (hence the static checks didn't run).

**Comment by teinarss** [comment]
Updated.

File: OpenRA.Game/OpenRA.Game.csproj [hunk 1]
```diff
@@ -25,7 +25,214 @@
   </PropertyGroup>
   <ItemGroup>
     <!-- Work around an issue where Rider does not detect files in the project root using the default glob -->
-    <Compile Include="**/*.cs" Exclude="$(DefaultItemExcludes)" />
```
Comment (reaperrr) [inline OpenRA.Game/OpenRA.Game.csproj]: Is this change necessary?

**Comment by teinarss** [comment]
Updated.

File: OpenRA.Game/Graphics/ChromeProvider.cs [hunk 2]
```diff
@@ -54,10 +54,10 @@ public class Collection
 
 		public static IReadOnlyDictionary<string, Collection> Collections { get; private set; }
 		static Dictionary<string, Collection> collections;
-		static Dictionary<string, Pair<Sheet, int>> cachedSheets;
+		static Dictionary<string, (Sheet Sheet, int density)> cachedSheets;
```
Comment (phrohdoh) [inline OpenRA.Game/Graphics/ChromeProvider.cs]: If I understand correctly that these are being treated as public members of an instance so are in PascalCase, I think `density` wants to be `Density`:
```suggestion
		static Dictionary<string, (Sheet Sheet, int Density)> cachedSheets;
```

**Review by abcdefg30 — DISMISSED** [review]
Lgtm.

**Comment by pchote** [comment]
I haven't looked at the code or tested this yet, but :+1: to the idea. A good followup PR will be to introduce `nameof` in the various places where we currently hardcode (sometimes incorrectly) strings for variable names, such as `[SequenceReference]` and `ArgumentException`.

It would be good if we could preemptively update the coding standard to ban the wackier C# features that have been introduced.

**Comment by teinarss** [comment]
> 
> 
> I haven't looked at the code or tested this yet, but 👍 to the idea. A good followup PR will be to introduce `nameof` in the various places where we currently hardcode (sometimes incorrectly) strings for variable names, such as `[SequenceReference]` and `ArgumentException`.
> 
> It would be good if we could preemptively update the coding standard to ban the wackier C# features that have been introduced.

Which ones did you have in mind?

**Comment by pchote** [comment]
The main things I don't like the idea of are expression bodied functions, pattern matching, and `using static`. I like string interpolation for variable names, but encoding arbitrary expressions/method calls inside strings is awful. I can also see the value of `?.` and local functions in specific situations, but IMO they discourage clean code design when used more generally.

**Comment by teinarss** [comment]
> 
> 
> The main things I don't like the idea of are expression bodied functions, pattern matching, and `using static`. I like string interpolation for variable names, but encoding arbitrary expressions/method calls inside strings is awful. I can also see the value of `?.` and local functions in specific situations, but IMO they discourage clean code design when used more generally.

* expression bodied functions - IMO these makes the code cleaner when having functions with only one line. For example in it the `CellLayer` class we have several methods that only have one line of code that code be changed to expression body style and would remove some of the clutter.
* using static - Totally agree.
*  I like string interpolation for variable names, but encoding arbitrary expressions/method calls inside strings is awful. - In these cases i would recommend us to extract that part to a variable.  If we are not already doing this for the `.F()` i would say that the same statement you said would apply on this code.
* `?.` and local functions in specific situations - `?.` is more useful and we have several places in the code base that we replace with this syntax, one of these is when we are using `events`.

File: OpenRA.Game/Graphics/SpriteFont.cs [hunk 3]
```diff
@@ -241,9 +241,9 @@ public int2 Measure(string text)
 			return new int2((int)Math.Ceiling(lines.Max(lineWidth)), lines.Length * size);
 		}
 
-		GlyphInfo CreateGlyph(Pair<char, Color> c)
+		GlyphInfo CreateGlyph((char c, Color color) c)
```
Comment (pchote) [inline OpenRA.Game/Graphics/SpriteFont.cs]: Most other places capitalize the tuple members, so shouldn't this be `(char C, Color Color)`?

File: OpenRA.Game/InstalledMods.cs [hunk 4]
```diff
@@ -34,9 +34,9 @@ public InstalledMods(IEnumerable<string> searchPaths, IEnumerable<string> explic
 			mods = GetInstalledMods(searchPaths, explicitPaths);
 		}
 
-		static IEnumerable<Pair<string, string>> GetCandidateMods(IEnumerable<string> searchPaths)
+		static IEnumerable<(string Name, string FullName)> GetCandidateMods(IEnumerable<string> searchPaths)
```
Comment (pchote) [inline OpenRA.Game/InstalledMods.cs]: I believe this should be `(string Id, string Path)` based on `Manifest LoadMod(string id, string path)`.

File: OpenRA.Game/Map/Map.cs [hunk 5]
```diff
@@ -729,7 +729,7 @@ public byte[] SavePreview()
 				var stride = bitmapWidth * 4;
 				var pxStride = 4;
 				var minimapData = new byte[stride * height];
-				Pair<Color, Color> terrainColor = default(Pair<Color, Color>);
+				(Color Left, Color Right) terrainColor = default((Color, Color));
```
Comment (pchote) [inline OpenRA.Game/Map/Map.cs:732]: `(Color Left, Color Right) terrainColor = default;` is valid C# 7.3, right? Probably makes sense to save this for a followup PR that simplifies `default`s more generally, though...

File: OpenRA.Game/Map/Map.cs [hunk 6]
```diff
@@ -738,7 +738,7 @@ public byte[] SavePreview()
 						var uv = new MPos(x + Bounds.Left, y + Bounds.Top);
 
 						// FirstOrDefault will return a Pair(MPos.Zero, Color.Transparent) if positions is empty
```
Comment (pchote) [inline OpenRA.Game/Map/Map.cs]: This comment will need to be updated too!

File: OpenRA.Game/Map/MapPreview.cs [hunk 7]
```diff
@@ -89,7 +89,7 @@ class InnerData
 			public bool DefinesUnsafeCustomRules { get; private set; }
 			public bool RulesLoaded { get; private set; }
 
-			public void SetRulesetGenerator(ModData modData, Func<Pair<Ruleset, bool>> generator)
+			public void SetRulesetGenerator(ModData modData, Func<(Ruleset Ruleset, bool CustomRules)> generator)
```
Comment (pchote) [inline OpenRA.Game/Map/MapPreview.cs]: `(Ruleset Ruleset, bool DefinesUnsafeCustomRules)` or `(Ruleset Ruleset, bool Flagged)`?

File: OpenRA.Game/Network/OrderManager.cs [hunk 8]
```diff
@@ -115,7 +115,7 @@ public void TickImmediate()
 				Connection.SendImmediate(localImmediateOrders.Select(o => o.Serialize()));
 			localImmediateOrders.Clear();
 
-			var immediatePackets = new List<Pair<int, byte[]>>();
+			var immediatePackets = new List<(int ClientId, byte[] Package)>();
```
Comment (pchote) [inline OpenRA.Game/Network/OrderManager.cs]: `(int ClientId, byte[] Packet)`?

File: OpenRA.Game/Network/ReplayConnection.cs [hunk 9]
```diff
@@ -23,7 +22,7 @@ public sealed class ReplayConnection : IConnection
 		class Chunk
 		{
 			public int Frame;
-			public Pair<int, byte[]>[] Packets;
+			public (int ClientId, byte[] Package)[] Packets;
```
Comment (pchote) [inline OpenRA.Game/Network/ReplayConnection.cs]: `(int ClientId, byte[] Packet)`?

File: OpenRA.Game/Network/ReplayConnection.cs [hunk 10]
```diff
@@ -55,7 +54,7 @@ public ReplayConnection(string replayFilename)
 			// to avoid issues with all immediate orders being resolved on the first tick.
 			using (var rs = File.OpenRead(replayFilename))
 			{
-				var packets = new List<Pair<int, byte[]>>();
+				var packets = new List<(int ClientId, byte[] Package)>();
```
Comment (pchote) [inline OpenRA.Game/Network/ReplayConnection.cs]: `(int ClientId, byte[] Packet)`?

File: OpenRA.Game/PlayerDatabase.cs [hunk 11]
```diff
@@ -83,15 +83,15 @@ public PlayerBadge LoadBadge(MiniYaml yaml)
 			{
 				sheetBuilder = new SheetBuilder(SheetType.BGRA, CreateSheet);
 
-				iconCache = new Cache<Pair<PlayerBadge, int>, Sprite>(p =>
+				iconCache = new Cache<(PlayerBadge PlayerBadge, int Density), Sprite>(p =>
```
Comment (pchote) [inline OpenRA.Game/PlayerDatabase.cs]: `(PlayerBadge Badge, int Density)`?

File: OpenRA.Mods.Common/Graphics/UIModelRenderable.cs [hunk 12]
```diff
@@ -106,13 +106,13 @@ public void Render(WorldRenderer wr)
...
 			{
-				return Screen3DBounds(wr).First;
+				return Screen3DBounds(wr).Bounds;
 			}
 
 			static readonly uint[] CornerXIndex = { 0, 0, 0, 0, 3, 3, 3, 3 };
 			static readonly uint[] CornerYIndex = { 1, 1, 4, 4, 1, 1, 4, 4 };
 			static readonly uint[] CornerZIndex = { 2, 5, 2, 5, 2, 5, 2, 5 };
-			Pair<Rectangle, float2> Screen3DBounds(WorldRenderer wr)
+			(Rectangle Bounds, float2) Screen3DBounds(WorldRenderer wr)
```
Comment (pchote) [inline OpenRA.Mods.Common/Graphics/UIModelRenderable.cs]: `(Rectangle Bounds, float2 Z)`?

File: OpenRA.Mods.Common/Lint/CheckChromeHotkeys.cs [hunk 13]
```diff
@@ -64,7 +64,7 @@ public void Run(Action<string> emitError, Action<string> emitWarning, ModData mo
 			}
 		}
 
-		void CheckInner(ModData modData, string[] namedKeys, Pair<string, string>[] checkWidgetFields, Dictionary<string, List<string>> customLintMethods,
+		void CheckInner(ModData modData, string[] namedKeys, (string, string Key)[] checkWidgetFields, Dictionary<string, List<string>> customLintMethods,
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckChromeHotkeys.cs]: `(string Widget, string Field)`?

File: OpenRA.Mods.Common/Lint/CheckChromeHotkeys.cs [hunk 14]
```diff
@@ -111,10 +109,9 @@ public void Run(Action<string> emitError, Action<string> emitWarning, ModData mo
 						checkArgKeys.AddRange(type.GetCustomAttributes<ChromeLogicArgsHotkeys>(true).SelectMany(x => x.LogicArgKeys));
 					}
 
-					Hotkey unused;
 					foreach (var n in node.Value.Nodes)
 						if (checkArgKeys.Contains(n.Key))
-							if (!namedKeys.Contains(n.Value.Value) && !Hotkey.TryParse(n.Value.Value, out unused))
+							if (!namedKeys.Contains(n.Value.Value) && !Hotkey.TryParse(n.Value.Value, out var unused))
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckChromeHotkeys.cs:114]: `out _`? Or are we saving discards for another general followup?

File: OpenRA.Mods.Common/Pathfinder/PathGraph.cs [hunk 15]
```diff
@@ -92,8 +92,8 @@ sealed class PathGraph : IGraph<CellInfo>
 		readonly bool checkTerrainHeight;
 		CellLayer<CellInfo> groundInfo;
 
-		readonly Dictionary<byte, Pair<ICustomMovementLayer, CellLayer<CellInfo>>> customLayerInfo =
-			new Dictionary<byte, Pair<ICustomMovementLayer, CellLayer<CellInfo>>>();
+		readonly Dictionary<byte, (ICustomMovementLayer CustomMovementLayer, CellLayer<CellInfo> CellInfos)> customLayerInfo =
```
Comment (pchote) [inline OpenRA.Mods.Common/Pathfinder/PathGraph.cs]: `(ICustomMovementLayer Layer, CellLayer<CellInfo> Info)`?

File: OpenRA.Mods.Common/Traits/Crates/Crate.cs [hunk 16]
```diff
@@ -155,20 +155,20 @@ void OnCrushInner(Actor crusher)
 
 			if (crateActions.Any())
 			{
-				var shares = crateActions.Select(a => Pair.New(a, a.GetSelectionSharesOuter(crusher)));
+				var shares = crateActions.Select(a => (CrateAction: a, Shares: a.GetSelectionSharesOuter(crusher)));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Crates/Crate.cs]: `(Action: , Shares: )`?

File: OpenRA.Mods.Common/Traits/GainsExperience.cs [hunk 17]
```diff
@@ -58,7 +58,7 @@ public class GainsExperience : INotifyCreated, ISync, IResolveOrder, ITransformA
 		readonly GainsExperienceInfo info;
 		readonly int initialExperience;
 
-		readonly List<Pair<int, string>> nextLevel = new List<Pair<int, string>>();
+		readonly List<(int XPRequirement, string Conditon)> nextLevel = new List<(int, string)>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/GainsExperience.cs]: `(int RequiredExperience, string Condition)`? (note the spelling fix!)

File: OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs [hunk 18]
```diff
@@ -97,7 +97,7 @@ public override void Activate(Actor self, Order order, SupportPowerManager manag
 			SendParatroopers(self, order.Target.CenterPosition, facing);
 		}
 
-		public Pair<Actor[], Actor[]> SendParatroopers(Actor self, WPos target, WAngle? facing = null)
+		public (Actor[] Aircrafts, Actor[] Units) SendParatroopers(Actor self, WPos target, WAngle? facing = null)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs]: `(Actor[] Aircraft, Actor[] Units)`?

File: OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs [hunk 19]
```diff
@@ -162,15 +162,15 @@ public SlotDropDownOption(string title, string order, Func<bool> selected)
 		}
 
 		/// <summary>Splits a string into two parts on the first instance of a given token.</summary>
-		static Pair<string, string> SplitOnFirstToken(string input, string token = "\\n")
+		static (string Text, string Desc) SplitOnFirstToken(string input, string token = "\\n")
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs]: `(string First, string Second)` or `(string Before, string After)`?

File: OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs [hunk 20]
```diff
@@ -253,9 +253,9 @@ public static void SelectSpawnPoint(OrderManager orderManager, MapPreviewWidget
 
 			var spawnSize = ChromeProvider.GetImage("lobby-bits", "spawn-unclaimed").Size.XY;
 			var selectedSpawn = preview.SpawnPoints
-				.Select((sp, i) => Pair.New(mapPreview.ConvertToPreview(sp, preview.GridType), i))
-				.Where(a => ((a.First - mi.Location).ToFloat2() / spawnSize * 2).LengthSquared <= 1)
-				.Select(a => a.Second + 1)
+				.Select((sp, i) => (MapRect: mapPreview.ConvertToPreview(sp, preview.GridType), Index: i))
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs]: `(SpawnLocation: , Index: )`?

File: OpenRA.Mods.Common/Widgets/Logic/SystemInfoPromptLogic.cs [hunk 21]
```diff
@@ -26,22 +25,22 @@ public class SystemInfoPromptLogic : ChromeLogic
 		// Increment the version number when adding new stats
 		const int SystemInformationVersion = 4;
 
-		static Dictionary<string, Pair<string, string>> GetSystemInformation()
+		static Dictionary<string, (string Text, string Value)> GetSystemInformation()
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/SystemInfoPromptLogic.cs]: `(string Label, string Value)`?

File: OpenRA.Mods.Common/Widgets/WidgetUtils.cs [hunk 22]
```diff
@@ -337,19 +337,19 @@ public static void BindPlayerNameAndStatus(LabelWidget label, Player p)
 		{
 			var client = p.World.LobbyInfo.ClientWithIndex(p.ClientIndex);
 			var nameFont = Game.Renderer.Fonts[label.Font];
-			var name = new CachedTransform<Tuple<string, WinState, Session.ClientState>, string>(c =>
+			var name = new CachedTransform<(string Text, WinState WinState, Session.ClientState ClientState), string>(c =>
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/WidgetUtils.cs]: `(string Name, WinState WinState, Session.ClientState ClientState)`?

**Review by pchote — COMMENTED** [review]
Code generally looks good to me. I have a few improved naming suggestions and other questions/comments:

File: OpenRA.Game/Map/Map.cs (hunk 5 above)
Comment (teinarss) [reply OpenRA.Game/Map/Map.cs:732]: I think so!

File: OpenRA.Mods.Common/Lint/CheckChromeHotkeys.cs (hunk 14 above)
Comment (teinarss) [reply OpenRA.Mods.Common/Lint/CheckChromeHotkeys.cs:114]: Yes

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
This is a big deal from a dev perspective, but doesn't have any user-facing impact so skipping the changelog entry for now.
If we (hopefully) get #17989 in then this can be added to the same entry.

I added a new line to the coding standard:
> Most code is written to the C# 5 feature set, but newer features are being adopted on a case-by-case basis. Prefer consistency with the surrounding code over always using the latest features, unless they provide meaningful benefit to your use case.

**Comment by pchote** [comment]
Created a changelog entry for the C# version update PRs: https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ffbf30f1775cc0796185234453e1be79d4353c83
